### DISCUSSION TITLE
How to create a JSON schema?

### DISCUSSION THREAD

**Discussion by Zamiell** [sinclairzx81/typebox#645] [Q&A]
Hello! I am interested in using TypeBox. I've read the README file, but it doesn't explain how to actually create a JSON schema. Is that documented somewhere?

For reference, I've been using `ts-json-schema-generator`, and in my build script I do this:

```
npx ts-json-schema-generator --path "./src/interfaces/IsaacScriptTSConfig.ts" --tsconfig "./tsconfig.json" --out "./schemas/tsconfig-isaacscript-section-schema.json"
```

So what is the analogous command in TypeBox land?

A separate question: Why doesn't this package use the `typebox` name? It would make installing it a lot easier. I see that [Sinclair seems to own the name on npm](https://www.npmjs.com/package/typebox), so there must be some other issue?

**Comment by sinclairzx81** [comment]
@Zamiell Hi,

---

> Hello! I am interested in using TypeBox. I've read the README file, but it doesn't explain how to actually create a JSON schema. Is that documented somewhere?

You can find documentation on creating Json Schema [here](https://github.com/sinclairzx81/typebox#types-json). 

All TypeBox types are Json Schema so you don't need to do anything else to create schematics (you just create types). Each `Type` function returns a small Json Schema fragment that combines into larger schematics.  For example, consider the following

```typescript
const N = Type.Number()              // const N = { type: 'number' }

const T = Type.Object({              // const T = {
  x: Type.Number(),                  //   type: 'object',
  y: Type.Number(),                  //   required: ['x', 'y', 'z'],
  z: Type.Number()                   //   properties: {
})                                   //     x: { type: 'number' },
                                     //     y: { type: 'number' },
                                     //     z: { type: 'number' }
                                     //   }
                                     // }
                                     //
console.log(T)                       // If you log T, it's just Json Schema
```
---

> For reference, I've been using ts-json-schema-generator, and in my build script I do this: .... So what is the analogous command in TypeBox land?

TypeBox isn't a code generator so it has no CLI commands to run. You just write types (as above) and pass them to validators to check.

```typescript
import { Type } from '@sinclair/typebox'
import Ajv from 'ajv'

const T = Type.Tuple([Type.Number(), Type.Number()])

const ajv = new Ajv()
ajv.validate(T, [1])          // false
ajv.validate(T, [1, 2])       // true
ajv.validate(T, [1, 2, 3])    // false
```
---

### Code Generation

If you need code generation, there are external community projects available that enable you to convert TypeScript to TypeBox, or Json Schema to TypeBox. See below for relevant links.

#### Json Schema to TypeBox

Code Generator https://github.com/xddq/schema2typebox

#### TypeScript to TypeBox

Code Generator https://github.com/sinclairzx81/typebox-codegen

Interactive https://sinclairzx81.github.io/typebox-workbench/


Hope this helps
S

**Reply by sinclairzx81** [reply]
Forgot this

> A separate question: Why doesn't this package use the typebox name? It would make installing it a lot easier. I see that [Sinclair seems to own the name on npm](https://www.npmjs.com/package/typebox), so there must be some other issue?

I will be publishing TypeBox 1.0 to `typebox`. For now, the recommendation is to use `@sinclair/typebox`.

**Reply by Zamiell** [reply]
Thank you!

Ok, sorry, what I wasn't understanding is that instead of having to run some CLI tool, I can just run `node "console.log(JSON.stringify(T))" > mySchema.json` to get the schema, which makes TypeBox different from e.g. Zod and all the other validators out there.

**Comment by woss** [comment]
for the people who stumble upon this, if you use fastify you can just return the schema object and it will be created as the json schema
